byzcoin/service.go

```diff
@@ -1802,7 +1831,16 @@ func (s *Service) getTxs(leader *network.ServerIdentity, roster *onet.Roster, sc
 	defer s.working.Done()
 	actualLeader, err := s.getLeader(scID)
 	if err != nil {
-		log.Lvlf1("%s: could not find a leader on %x with error: %s", s.ServerIdentity(), scID, err)
+		log.Lvlf2("%s: could not find a leader on %x with error: %s", s.ServerIdentity(), scID, err)
+		if s.skService().ChainIsFriendly(scID) {
+			log.Lvlf1("%s: catching up with chain %x", s.ServerIdentity(), scID)
+			err = s.catchupFromID(roster, scID)
```

Comment:
IIRC, this code is to have an additional trigger to download the chain, in case the node started up with a fresh database, and is then contacted by the leader to give txs, but the new node doesn't even know it's part of this byzcoin instance. So we cannot take the latest roster, because we don't have it here.

On the other hand, this might be too agressive in trying to catch up. But what other possibilities does a node have if his database got corrupted, removed, and now he has to start fresh? He could wait for the leader to be asked to sign a block. But in the current usage, most of the time the nodes are idle, and not signing blocks. So the only message we get from the leader is this `collectTx`.

Furthermore, `catchupID` starts by doing `GetUpdateChain`, which, as we saw earlier, does a trusted update of the chain. So if a wrong leader tries to do something sneaky, we'll get to know it in there, and will simply error out.

So using https://github.com/dedis/cothority/blob/master/scmgr/app.go#L175 gets more important. This is something I did on the latest nodes for personhood, and it seems to work fine. It avoids me using the nativescript-tests to add a new byzcoin if I forget to change the roster for the tests ;)

I vote for keeping it like that.